File: LeetCode/graph/797.all-paths-from-source-to-target.py
#
# @lc app=leetcode id=797 lang=python3
#
# [797] All Paths From Source to Target
#

# @lc code=start
class Solution:
    result = []
    def traverse(self, graph, lastNodeIdx, rootNode, myRoute):
        # rootNode is actually index
        myRoute.append(rootNode)
        if rootNode == lastNodeIdx:
            self.result.append(myRoute.copy()) 
            myRoute.pop(-1)
            return 
        for node in graph[rootNode]:  # nodes in rootNode.neighbors
            self.traverse(graph, lastNodeIdx, node, myRoute)
        myRoute.pop(-1)
               
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        self.result = []
        lastNodeIdx = len(graph)-1
        self.traverse(graph, lastNodeIdx, 0, [])
        return self.result


# @lc code=end

# 这道题的关键是不使用visited, 从0开始对图进行遍历       
# 因为不使用visited， 在对每一个点进行遍历的时候他们都尝试了所有可能的选择
# 可以使用preVisit(), postVisit()思想动态维护list myRoute, 节约.copy()的花费

# with copy()
    def traverse(self, graph, lastNodeIdx, rootNode, myRoute):
        # rootNode is actually index
        myRoute.append(rootNode)
        print(myRoute)
        if rootNode == lastNodeIdx: 
            self.result.append(myRoute) 
            return 
        for node in graph[rootNode]:  # nodes in rootNode.neighbors
            self.traverse(graph, lastNodeIdx, node, myRoute.copy())

File: LeetCode_Go/Graph/797.all-paths-from-source-to-target.go
/*
 * @lc app=leetcode id=797 lang=golang
 *
 * [797] All Paths From Source to Target
 */

// @lc code=start

// Given a directed acyclic graph (DAG) of n nodes labeled from 0 to n - 1 (adjacent list), find all possible paths from node 0 to node n - 1 and return them in any order (return array of array(path)).

// DFS traverse logic, don't use visited array, append to path when enter an node, and undo appending path when finished the node. Record path when reach target.

func allPathsSourceTarget(graph [][]int) [][]int {
	ans := [][]int{}
	path := []int{}
	DFS(graph, 0, &path, &ans)
	return ans
}

func DFS(graph [][]int, v int, path *[]int, ans *[][]int) {
	*path = append(*path, v)
	// reach last node, record path
	if v == len(graph)-1 {
		temp := make([]int, len((*path)))
		copy(temp, *path)
		*ans = append(*ans, temp)
	}

	for _, i := range graph[v] {
		// fmt.Printf("%v is calling DFS(%v)\n", *path, i)
		DFS(graph, i, path, ans)
	}

	*path = (*path)[:len(*path)-1]
}

// @lc code=end

